File: Ruby/sample.rb
require 'rubygems'
require 'json'
require 'net/https'
require 'net/http'


module Mixi
  CONSUMER_KEY = "[YOUR CONSUMER KEY]"
  CONSUMER_SECRET = "[YOUR CONSUMER SECRET]"
  REDIRECT_URI = "[YOUR REDIRECT URI]"

  class Client
    def self.example
      raise 'authcode is mandatory.' if ARGV.empty?
      client = Mixi::Client.new
      client.authorize! ARGV.first
      client.call '/people/@me/@self'
    end

    def authorize!(auth_code)
      params = "grant_type=authorization_code&client_id=#{CONSUMER_KEY}&client_secret=#{
        CONSUMER_SECRET}&code=#{auth_code}&redirect_uri=#{REDIRECT_URI}"
      https = Net::HTTP.new('secure.mixi-platform.com', 443)
      https.use_ssl = true
      https.start do
        response = https.post('/2/token', params).body
        json = JSON.parse(response)
        @access_token = json['access_token']
        @refresh_token = json['refresh_token']
      end
    end

    def call(endpoint)
      http = Net::HTTP.new('api.mixi-platform.com', 80)
      http.start do
        response = http.get("/2#{endpoint}",
                            'Authorization' => "OAuth #{@access_token}",
                            'HOST' => 'api.mixi-platform.com').body
        JSON.parse(response)
      end
    end
  end
end

Mixi::Client.example

File: Perl/sample.pl
#!/usr/bin/perl

use strict;
use warnings;

use constant {
    CONSUMER_KEY    => '[YOUR CONSUMER KEY]',
    CONSUMER_SECRET => '[YOUR CONSUMER SECRET]',
    REDIRECT_URI    => '[YOUR REDIRECT URI]',

    MIXI_TOKEN_ENDPOINT => 'https://secure.mixi-platform.com/2/token',
    MIXI_API_ENDPOINT   => 'http://api.mixi-platform.com/2'
};

use URI::Escape qw/ uri_escape /;
use HTTP::Request;
use LWP::UserAgent;
use JSON::XS;
use Data::Dumper;

sub request {
    my ($method, $url, $data_arr) = @_;

    my $req = HTTP::Request->new(
        $method => $url
    );

    if ($method eq 'POST') {
        my $data_str = join '&', map { uri_escape($_) . '=' . uri_escape($data_arr->{$_}) } keys %$data_arr;
        $req->content_type('application/x-www-form-urlencoded');
        $req->content($data_str);
    }

    my $ua = LWP::UserAgent->new();
    my $res = $ua->request($req);

    return $res;
}

sub get_token {
    my $auth_code = shift;

    my %data_arr = (
        'grant_type'    => 'authorization_code',
        'client_id'     => CONSUMER_KEY,
        'client_secret' => CONSUMER_SECRET,
        'code'          => $auth_code,
        'redirect_uri'  => REDIRECT_URI,
    );

    my $res = request('POST', MIXI_TOKEN_ENDPOINT, \%data_arr);
    die 'Request failed. ' . $res->status_line unless $res->is_success;

    return decode_json($res->content);
}

sub get_new_token {
    my $refresh_token = shift;

    my %data_arr = (
        'grant_type'    => 'refresh_token',
        'client_id'     => CONSUMER_KEY,
        'client_secret' => CONSUMER_SECRET,
        'refresh_token' => $refresh_token,
    );

    my $res = request('POST', MIXI_TOKEN_ENDPOINT, \%data_arr);
    die 'Request failed. ' . $res->status_line unless $res->is_success;

    return decode_json($res->content);
}

sub call {
    my ($endpoint, $token) = @_;

    my $url = MIXI_API_ENDPOINT . $endpoint . '?oauth_token=' . uri_escape($token->{'access_token'});
    my $res = request('GET', $url);

    if (defined $res->header('WWW-Authenticate')) {
        my $error_msg = $res->header('WWW-Authenticate');

        if ($error_msg =~ /invalid_request/) {
            die 'Invalid request.';
        } elsif ($error_msg =~ /invalid_token/) {
            die 'Invalid token.';
        } elsif ($error_msg =~ /expired_token/) {
            $token = get_new_token($token->{'refresh_token'});
            $url = MIXI_API_ENDPOINT . $endpoint . '?oauth_token=' . uri_escape($token->{'access_token'});
            return decode_json(request('GET', $url)->{'_content'});
        } elsif ($error_msg =~ /insufficient_scope/) {
            die 'Insufficient scope.';
        }
    }

    return decode_json($res->content);
}

my $auth_code = $ARGV[0];
my $token = get_token($auth_code);
my $json_href = call('/people/@me/@self', $token);

print Data::Dumper::Dumper($json_href);

1;

